activerecord/lib/active_record/migration.rb

```diff
@@ -553,21 +553,40 @@ def self.current_version
     # This class is used to verify that all migrations have been run before
     # loading a web page if <tt>config.active_record.migration_error</tt> is set to :page_load
     class CheckPending
-      def initialize(app)
+      def initialize(app, file_watcher: ActiveSupport::FileUpdateChecker)
         @app = app
-        @last_check = 0
+        @needs_check = true
+        @mutex = Mutex.new
+        @file_watcher = file_watcher
       end
 
       def call(env)
-        mtime = ActiveRecord::Base.connection.migration_context.last_migration.mtime.to_i
-        if @last_check < mtime
-          ActiveRecord::Migration.check_pending!(connection)
-          @last_check = mtime
+        @mutex.synchronize do
+          @watcher ||= @file_watcher.new(*watchable_args) do
```

Comment:
Would this be clearer as something like `new([], existing_migration_files)`?

Reply:
I agree something feels improvable here.

The second argument is a hash in the format `{ directory => [extensions] }`, so calling it `_files` or `_dirs` feels off (I'd expect that to be an array). `watchable_dirs_arg` is descriptive, but not much cleaner.

I decided to rename the other method `build_watcher`, have it take a block and return the file watcher instance. I think that is clearer ❤️